lock_api/src/mutex.rs

```diff
@@ -780,10 +788,12 @@ impl<R: RawMutex, T: ?Sized> Drop for ArcMutexGuard<R, T> {
 /// could introduce soundness issues if the locked object is modified by another
 /// thread.
 #[must_use = "if unused the Mutex will immediately unlock"]
-pub struct MappedMutexGuard<'a, R: RawMutex, T: ?Sized> {
+pub struct MappedMutexGuard<'a, R: RawMutex, T: ?Sized + 'a> {
     raw: &'a R,
-    data: *mut T,
-    marker: PhantomData<&'a mut T>,
+    // We use `&'a mut` to make this type invariant over `'a`
+    marker: PhantomData<&'a mut ()>,
```

Comment:
Does it actually need to be invariant here? I think we can safely remove this.

Reply:
I kept it invariant because that was how it was in the original code, and it's the most conservative option. Without it this type would be variant, which is probably correct but there's a chance of it being subtly broken.

Reply:
Actually this made `T` invariant, not `'a` (see https://doc.rust-lang.org/nomicon/subtyping.html#variance).

The proper way to get invariance on `'a` is  `PhantomData<fn(&'a ()) -> &'a ()>`